### PR TITLE
wrap models using macros

### PR DESCRIPTION
**Pull request by drewbanin**
- this removes the `Model.compile` call (and the associated, duplicate context that we need to create).

This branch is currently a proof-of concept, but in it's current state:

- macros work in hooks
- `already_exists` works (kind of)
- we no longer "compile twice" (we do in practice, but we won't have to with this setup)

### CODE REVIEW

File: dbt/adapters/redshift.py [hunk 1]
```diff
@@ -30,7 +31,7 @@ def sort_qualifier(cls, sort_type, sort):
                 .format(sort_type, valid_sort_types)
             )
 
-        if type(sort) == str:
```
Comment (cmcarthur) [inline dbt/adapters/redshift.py]: great catch

File: dbt/include/global_project/macros/materializations/view.sql [hunk 2]
```diff
@@ -0,0 +1,17 @@
+{% macro dbt__create_view(schema, model, sql, flags, funcs) -%}
+
+  {%- set identifier = model['name'] -%}
+  {%- set already_exists = funcs.already_exists(schema, identifier) -%}
+  {%- set non_destructive_mode = flags.NON_DESTRUCTIVE == True -%}
+
+  {%- if non_destructive_mode -%}
```
Comment (drewbanin) [inline dbt/include/global_project/macros/materializations/view.sql]: is this the right logic for views in non-destructive mode?

File: dbt/parser.py [hunk 3]
```diff
@@ -142,7 +188,7 @@ def parse_node(node, node_path, root_project_config, package_project_config,
 
     context = {}
 
-    context['ref'] = lambda *args: ''
+    context['ref'] = __ref(node)
```
Comment (drewbanin) [inline dbt/parser.py]: +1 +1 +!

File: dbt/runner.py [hunk 4]
```diff
@@ -239,6 +239,9 @@ def execute_model(profile, model, existing):
 
     result = None
 
+    if get_materialization(model) == 'ephemeral':
+        return 'SKIP'
```
Comment (drewbanin) [inline dbt/runner.py]: we're just going to silently skip these, right?

File: dbt/wrapper.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    if 'sort' not in model['config']:
+        return ''
+
+    if get_materialization(model) not in ('table', 'incremental'):
+        return ''
+
+    sort_keys = model_config.get('sort')
+    sort_type = model_config.get('sort_type', 'compound')
+
+    if type(sort_type) != str:
```
Comment (drewbanin) [inline dbt/wrapper.py]: :)

File: dbt/parser.py [hunk 6]
```diff
@@ -399,7 +396,7 @@ def parse_schema_test(test_base, model_name, test_config, test_type,
 
     raw_sql = "{{{{ {macro}(model=ref('{model}'), {kwargs}) }}}}".format(**{
         'model': model_name,
-        'macro': "test_{}".format(test_type),
+        'macro': "dbt.test_{}".format(test_type),
```
Comment (drewbanin) [inline dbt/parser.py]: this is ok for now, but i think ideally we wouldn't namespace this. Idea being that end users could 1) define their own tests (not in the dbt namespace) or 2) override existing tests

Comment (cmcarthur) [inline dbt/parser.py]: good point, i will switch this back.

**Pull request merged**